src/store/useRoslib.ts

```diff
@@ -0,0 +1,219 @@
+import { defineStore } from 'pinia';
+import ROSLIB from 'roslib';
+import { ref } from 'vue';
+import type { Ref } from 'vue';
+import type { TopicType, StdMsg, TopicTypeMap, RosCompressionType } from './rosTypes';
+
+const HEARTBEAT_DISCONNECT_SECONDS = 2;
+const RECONNECTION_GRACE_SECONDS = 2;
+const SECONDS_TO_TIMESTAMP = 1000;
+
+export const useRoslibStore = defineStore('roslib', () => {
+  const ros = new ROSLIB.Ros({ url: undefined });
+  const isWebSocketConnected = ref<boolean>(false);
+  const stop = ref<boolean>(false);
+  function init(serverHost: string) {
+    ros.connect(serverHost);
+    console.count('Reinitialized RosInit');
+    // TODO: Realistically, we should use websocket ping/pong control frames for this.
+    //       This potentially requires hooking into rosbridge_server, so we'll leave it for the future.
+    //       With this, we can increase the heartbeat frequency.
+    let heartbeatTime: number = Date.now() + RECONNECTION_GRACE_SECONDS * SECONDS_TO_TIMESTAMP;
+
+    ros.on('connection', () => {
+      isWebSocketConnected.value = true;
+    });
+    ros.on('error', (error) => {
+      isWebSocketConnected.value = false;
+      console.error(error);
+    });
+    ros.on('close', () => {
+      isWebSocketConnected.value = false;
+    });
+    // TODO: refactor to retrieve time instead of boolean
+    const [isReceivingHeartBeatData, heartbeatSub] = createSubscriber({
+      topicName: 'hbr',
+      topicType: 'std_msgs/Bool',
+      startingDefaultValue: false,
+    });
+    heartbeatSub({ defaultValue: false, isDebugging: true });
+    // Close the connection if the heartbeat stops for too long.
+    const interval = setInterval(() => {
+      if (isReceivingHeartBeatData.value) {
+        heartbeatTime = Date.now();
+      }
+      if (stop.value) {
+        clearInterval(interval);
+
+        // No need to error if the websocket is already closed.
+        try {
+          ros.close();
+        } catch (_) {
+          /* empty */
+        }
+
+        return;
+      }
+      if (Date.now() - heartbeatTime > HEARTBEAT_DISCONNECT_SECONDS * SECONDS_TO_TIMESTAMP) {
+        isWebSocketConnected.value = false;
+
+        // No need to error if the websocket is already closed.
+        try {
+          ros.close();
+        } catch (_) {
+          /* empty */
+        }
+
+        // Forcefully reconnect.
+        // @ts-expect-error socket is private but needs to be set to null to force roslib to reconnect in case of hangs.
+        ros.socket = null;
+        // Give the reconnection extra time so that it doesn't
+        // immediately get killed.
+        heartbeatTime = Date.now() + RECONNECTION_GRACE_SECONDS * SECONDS_TO_TIMESTAMP;
+        ros.connect(serverHost);
+      }
+    }, 100);
+  }
+  function createTopic<T>(
+    topicName: string,
+    topicType: TopicType,
+    compression: RosCompressionType = 'cbor',
+  ) {
+    return new ROSLIB.Topic<StdMsg<T>>({
```

Comment:
note that not all msg types will have a data field; it would be better to pass `T` into `Topic` and make the caller use the `StdMsg<T>` type.